components/common/src/templating/mod.rs

```diff
@@ -91,6 +95,35 @@ impl TemplateRenderer {
             .render(template, &raw)
             .map_err(|e| Error::TemplateRenderError(format!("{}", e)))
     }
+
+    // This method is only implemented so we can intercept the call to Handlebars and display
+    // a deprecation message to users. More information here https://github.com/habitat-sh/habitat/issues/6323.
+    // When Handlebars is upgraded this can be safely removed.
+    pub fn register_template_file(&mut self,
+                                  name: &str,
+                                  path: &std::path::Path)
+                                  -> result::Result<(), TemplateFileError> {
+        let re = Regex::new(r"\{\{[^}]+[^.]\[[^}]*\}\}").expect("Failed to compile template \
+                                                                 deprecation regex");
+        if let Ok(template_string) = read_to_string(path) {
+            if re.is_match(&template_string) {
+                println!("\n\n********** Deprecation Notice!!! **********\n\nThe template at \
+                          {:?} is using non-standard handlebars syntax for array access. \
+                          Discussed here https://github.com/habitat-sh/habitat/issues/6323. \
+                          Support for this syntax will be removed in an upcoming \
+                          release.\n\nNon-standard syntax found in the following statements:\n",
+                         path);
+
+                for mat in re.find_iter(&template_string) {
+                    println!("{}", mat.as_str());
```

Comment:
I would be more user-friendly to provide a line/column location of the deprecated syntax as well as a suggestion for how to correct it. The former might require splitting `template_string` into lines first since [the `Match` struct](https://docs.rs/regex/1.1.5/regex/struct.Match.html) only gives byte offsets.

Reply:
We *might* even be able to do the replacement with good syntax on the fly, allowing us to do the upgrade to the new version of handlebars now.

Reply:
I like the idea of updating the syntax on the fly; that feels like something we should do when we actually do the upgrade though.

Reply:
Sure, there's really no point in doing so *until* we upgrade our handlebars library. What I'm saying is that instead of:
1. Display a deprecation warning for some number of releases, hoping that people fix their templates 
1. Update to the new handlebars library and break any un-fixed templates

we can:
1. Display a deprecation warning for some number of releases, hoping that people fix their templates 
1. Update to the new handlebars library, fixing instances of the old syntax on the fly (with a warning)
1. Optionally at some point in the future remove the fixup code altogether and break any un-fixed templates

Previously we thought we were *unable* to update our very old version of handlebars because we didn't want to break people's templates that use the old syntax. The point is, given how simply we could fix up the old syntax on the fly, I don't think we need to wait on updating the handlebars library.

I do think starting with just a deprecation warning is a good idea, since there could be bugs with our on-the-fly fixes, but I also think there's a lot of value in updating to the new handlebars library version sooner rather than later.

Reply:
Yup, I think we are agreeing with each other.